core/modules/qana/QservRestrictorPlugin.cc

```diff
@@ -571,14 +574,36 @@ QservRestrictorPlugin::_newSecIndexRestrictor(
         return query::QsRestrictor::Ptr();
     }
 
+    // Get the director table (or an empty string)
+    std::string dirTable = context.cssFacade->getDirTable(cr->db, cr->table);
+
     // Build the QsRestrictor
     query::QsRestrictor::Ptr p = std::make_shared<query::QsRestrictor>();
     p->_name = "sIndex";
     // sIndex has paramers as follows:
     // db, table, column, val1, val2, ...
-    p->_params.push_back(cr->db);
-    p->_params.push_back(cr->table);
-    p->_params.push_back(cr->column);
+    if (!dirTable.empty()) {
+        std::string dirDb = context.cssFacade->getDirDb(cr->db, cr->table);
+        if (dirDb.empty()) {
+            dirDb = cr->db;
+        }
+        std::string dirCol = context.cssFacade->getDirColName(cr->db, cr->table);
+        if (dirCol.empty()) {
+            LOGF_ERROR("dirTable set, dirCol missing for %1%.%2%" % cr->db % cr->table);
+            return query::QsRestrictor::Ptr();
+        }
+        LOGF_DEBUG("Using dirDb %1%, dirTable %2%, dirCol %3% as sIndex for %4%.%5%." %
+                   dirDb % dirTable % dirCol % cr->db % cr->table);
+        p->_params.push_back(dirDb);
+        p->_params.push_back(dirTable);
+        p->_params.push_back(dirCol);
+    } else {
+        LOGF_DEBUG("Using %1%, %2%, %3% as sIndex." %
+                   cr->db % cr->table % cr->column);
+        p->_params.push_back(cr->db);
+        p->_params.push_back(cr->table);
+        p->_params.push_back(cr->column);
+    }
```

Comment:
No.

The first time you set dirCol, you set it to the name of the column **for the current table**. You have to set it again to the name of the column **in the director table**, once you know what the director table is (and that it is not equal to the current table). Consider for example a Source table with a director column named "directorId" that corresponds to the "objectId" column in a director table named Object.

As far as whether dirColName should be allowed without a dirTable - no idea. I thought it might be reasonable to default to the current table if no director table is explicitly specified. If we should be checking that dirTable and dirColName are always set together, it's a very small mod to the initial proposed snippet. FWIW I find your sequence of if statements very confusing, you are using `or` (non-idiomatic C++) instead of `||`, and the final error when cr->col != dirCol is not correct.